### PR TITLE
Adapt  and implement `DELEG` STS rule properties

### PR DESCRIPTION
**Pull request by mgudemann**
Implement properties for checking delegation certificates used in the `DELEG` STS rule.

Checks that:
 * Credential deregistration correctly removes the credential from the mappings
 * Credential delegation correctly changes the delegation mappings

### CODE REVIEW

File: shelley/chain-and-ledger/executable-spec/test/Rules/TestDeleg.hs [hunk 1]
```diff
@@ -79,3 +80,42 @@ rewardZeroAfterReg = withTests (fromIntegral numberOfTests) . property $ do
...
+credentialMappingAfterDelegation :: Property
+credentialMappingAfterDelegation = withTests (fromIntegral numberOfTests) . property $ do
+  t <- forAll (trace @DELEG $ fromIntegral traceLen)
+  let
+    n :: Integer
+    n = fromIntegral $ traceLength t
+    tr = sourceSignalTargets t
+
+  when (n > 1) $
+    True === (all delegatedCredential tr)
```
Comment (dnadales) [inline shelley/chain-and-ledger/executable-spec/test/Rules/TestDeleg.hs]: Does it make sense to use `assert all ...`?

File: shelley/chain-and-ledger/executable-spec/test/Rules/TestDeleg.hs [hunk 2]
```diff
@@ -79,3 +80,42 @@ rewardZeroAfterReg = withTests (fromIntegral numberOfTests) . property $ do
...
+credentialRemovedAfterDereg :: Property
+credentialRemovedAfterDereg = withTests (fromIntegral numberOfTests) . property $ do
+  t <- forAll (trace @DELEG $ fromIntegral traceLen)
+  let
+    n :: Integer
+    n = fromIntegral $ traceLength t
+    tr = sourceSignalTargets t
+
+  when (n > 1) $
+    True === (all removedDeregCredential tr)
```
Comment (dnadales) [inline shelley/chain-and-ledger/executable-spec/test/Rules/TestDeleg.hs]: Maybe use `assert`?

Comment (dnadales) [reply shelley/chain-and-ledger/executable-spec/test/Rules/TestDeleg.hs]: I'm also thinking wouldn't it make sense to check that the set of non-removed delegation credential is empty? The difference here is that if the tests fail we could see which credentials didn't get removed.

File: shelley/chain-and-ledger/executable-spec/test/Rules/TestDeleg.hs (hunk 1 above)
Comment (dnadales) [reply shelley/chain-and-ledger/executable-spec/test/Rules/TestDeleg.hs]: Or like I wrote above, maybe we want to use a set here.

**Comment by mgudemann** [comment]
@dnadales thanks for the review, I made it an `assert all` which is much better of course.
For the rest, I tried to make it as close as possible to the way the properties are written in the spec at the moment.
But you're right of course, it makes sense to formulate that all other credentials stay registered. I'd just go with a separate property for this.

**Comment by dnadales** [comment]
> But you're right of course, it makes sense to formulate that all other credentials stay registered. I'd just go with a separate property for this.

Alright. However my comment was that we should replace the assertion (or `True === ...` check) by an `... /== Set.empty` check, so that when the tests fail we have more information than just `False`.

**Pull request merged**